File: api/src/console.cpp
#include "console.h"

#define MAX_DEPTH 2
#define MAX_TIME 5
#define DEBUG true
#define USE_HUMAN true

namespace coredump
{
    // 0 is valid, 1 is invalid move, -1 is quit
    int makeHumanTurn(const Position &currentPosition, Color currentPlayer, Move &move)
    {
        std::cout << "\nCommands:" << std::endl;
        std::cout << "- Make move: e2 e4" << std::endl;
        // std::cout << "- Show moves for square: moves e2" << std::endl;
        std::cout << "- Quit: quit> " << std::endl;
        std::string input = "";
        std::getline(std::cin, input);

        if (input == "quit")
        {
            std::cout << "Quitting the game" << std::endl;
            return -1;
        }
        // Handle move command
        if (input.length() != 5 || input[2] != ' ')
        {
            std::cout << "Invalid move format. Use 'e2 e4'\n";
            return 1;
        }

        // Create move to check legality
        move.fromSquare = Move::fromAlgebraic(input[0], input[1]);
        move.toSquare = Move::fromAlgebraic(input[3], input[4]);
        move.color = currentPlayer;

        std::vector<Move> legalMoves = generateMoves(currentPosition, currentPlayer);
        auto moveIt = std::find_if(legalMoves.begin(), legalMoves.end(),
                                   [&move](const Move &m)
                                   {
                                       return m.fromSquare == move.fromSquare && m.toSquare == move.toSquare;
                                   });

        if (moveIt == legalMoves.end())
        {
            std::cout << "Illegal move." << std::endl;
            return 1;
        }

        move = *moveIt; // Copy extra properties from legal move

        // Handle pawn promotion
        if (move.isPromotion)
        {
            std::cout << "Choose promotion piece Q(ueen), R(ook), B(ishop), k(N)ight: ";
            char choice;
            std::cin >> choice;
            std::cin.ignore(std::numeric_limits<std::streamsize>::max(), '\n');

            move.promotionPiece = getPromotionPiece(choice);
            if (move.promotionPiece == PieceType::NONE)
            {
                std::cout << "Invalid promotion piece. Defaulting to Queen.\n";
                move.promotionPiece = PieceType::QUEEN;
            }
        }

        return 0; // Valid move
    }

    int start_console()
    {
        // Initialize the position
        Position currentPosition;

        // TODO make this run more automatically for when in the pybind module
        initializeMagicBitboards();

        Color currentPlayer = Color::WHITE; // White moves first
        Color humanColor = Color::WHITE;    // Human plays white by default
        int fullmoveCounter = 0;
        int halfmoveClock = 0;
        std::ostringstream pgn;

        // Let player choose color
        std::cout << "Play as (w)hite or (b)lack? ";
        char choice;
        std::cin >> choice;
        std::cin.ignore(std::numeric_limits<std::streamsize>::max(), '\n');
        if (tolower(choice) == 'q')
        {
            std::cout << "Quitting the game" << std::endl;
            return 0;
        }
        if (tolower(choice) == 'b')
        {
            humanColor = Color::BLACK;
        }

        // print out who plays who
        std::cout << "You are playing as " << colorToString(humanColor) << std::endl;
        std::cout << "AI is playing as " << colorToString(invertColor(humanColor)) << std::endl;
        if (humanColor == Color::WHITE)
        {
            std::cout << "You play first" << std::endl;
        }
        else
        {
            std::cout << "AI plays first" << std::endl;
        }

        // Main game loop
        while (true)
        {
            // Display the current position
            std::cout << "\n====================" << std::endl;
            std::cout << "Move " << fullmoveCounter << std::endl;

            std::cout << colorToString(currentPlayer) << " to move\n";

            int status = checkEndgameConditions(currentPosition, currentPlayer);
            if (status == 1)
            {
                std::cout << "CHECK!\n";
            }
            else if (status == 2)
            {
                std::cout << "CHECKMATE! " << colorToString(invertColor(currentPlayer)) << " wins." << std::endl;
                break;
            }
            else if (status == 3)
            {
                std::cout << "STALEMATE! Nobody wins" << std::endl;
                break;
            }

            // TODO add more endgame conditions
            // else if (isInsufficientMaterial(currentPosition)) {
            // 	std::cout << "INSUFFICIENT MATERIAL!\n";
            // 	break;
            // }
            // else if (isThreefoldRepetition(currentPosition)) {
            // 	std::cout << "THREEFOLD REPETITION!\n";
            // 	break;
            // }
            // else if (isFiftyMoveRule(currentPosition)) {
            // 	std::cout << "FIFTY MOVE RULE!\n";
            // 	break;
            // }

            std::cout << currentPosition.displayPosition() << std::endl;
            std::cout << currentPosition.getFen(currentPlayer, halfmoveClock, fullmoveCounter, "", "") << std::endl;
            std::cout << "PGN:\n";
            std::cout << pgn.str() << std::endl;

            // Determine if it's human's turn
            bool isHumanTurn = (currentPlayer == humanColor) && USE_HUMAN;
            Move move;
            if (isHumanTurn)
            {
                bool quit = false;
                while (true)
                {
                    int status = makeHumanTurn(currentPosition, currentPlayer, move);
                    if (status == 0)
                    {
                        break; // Move was made successfully
                    }
                    else if (status == -1)
                    {
                        quit = true;
                        break;
                    }
                }
                if (quit)
                {
                    break;
                }
            }
            else
            {
                // AI's turn
                std::cout << "AI is thinking...\n";
                std::vector<Move> legalMoves = generateMoves(currentPosition, currentPlayer);

                std::cout << "Legal moves: " << legalMoves.size() << std::endl;
                // print all the legal moves in algebraic notation
                for (const auto &m : legalMoves)
                {
                    Position tempPos(currentPosition);
                    tempPos.makeMove(m);
                    int evaluation = evaluatePosition(tempPos, currentPlayer);
                    std::cout << Move::toAlgebraic(m.fromSquare) << " " << Move::toAlgebraic(m.toSquare) << " scores " << evaluation << std::endl;
                }

                move = findBestMove(currentPosition, currentPlayer, MAX_DEPTH, MAX_TIME, DEBUG);
                // Convert move to algebraic notation for display
                std::string moveStr = Move::toAlgebraic(move.fromSquare) + " " + Move::toAlgebraic(move.toSquare);
                std::cout << "Computer plays: " << moveStr << std::endl;
            }

            // Make the selected move
            currentPosition.makeMove(move);

            // update pgn
            if (currentPlayer == Color::WHITE)
            {
                fullmoveCounter += 1;
                pgn << fullmoveCounter;
                pgn << '.';
            }
            pgn << move.getPgn();
            pgn << ' ';

            // Switch to next player
            currentPlayer = invertColor(currentPlayer);
        }

        std::cout << "\nGame ended.\n";
        std::cout << "PGN:\n";
        std::cout << pgn.str() << std::endl;
        return 0;
    }
}

File: console.py
import build.core_dump_py as cd
from build.core_dump_py import Color, PieceType, Move, Position

# Initialize important setup, like magic bitboards
cd.engine_init()

MAX_DEPTH = 3
MAX_TIME = 5
DEBUG = False
USE_HUMAN = False
CHOOSE_RANDOMLY = False

def make_human_turn(current_position, current_player):
    print("\nCommands:")
    print("- Make move: e2 e4")
    # print("- Show moves for square: moves e2")
    print("- Quit: quit> ")
    input_str = input().strip()

    if input_str == "quit":
        print("Quitting the game")
        return -1, None

    # Handle move command
    if len(input_str) != 5 or input_str[2] != ' ':
        print("Invalid move format. Use 'e2 e4'")
        return 1, None

    # Create move to check legality
    move = Move(input_str, current_player)

    legal_moves = cd.generate_moves(current_position, current_player)
    legal_move = next((m for m in legal_moves if m.from_square == move.from_square and m.to_square == move.to_square), None)

    if legal_move is None:
        print("Illegal move.")
        return 1, None

    move = legal_move  # Copy extra properties from legal move

    # Handle pawn promotion
    if move.is_promotion:
        print("Choose promotion piece Q(ueen), R(ook), B(ishop), k(N)ight: ", end="")
        choice = input().strip().upper()
        pieceChoice = PieceType.NONE
        try:
            pieceChoice = cd.get_promotion_piece(choice)
            if pieceChoice == PieceType.NONE:
                raise ValueError
        except ValueError:
            print("Invalid promotion piece. Defaulting to Queen.")
            pieceChoice = PieceType.QUEEN
        finally:
            print(f"Promoting to {pieceChoice}")
            move.set_promotion_piece(pieceChoice)

    return 0, move  # Valid move

# result (0 for continue, -1 for quit), current_position, current_player, full_move_counter, halfmove_clock, pgn
def game_loop(current_position, current_player, human_color, full_move_counter, halfmove_clock, pgn):
    print("\n====================")
    print(f"Move {full_move_counter}, {current_player.to_string()} to move")

    status = cd.check_endgame_conditions(current_position, current_player)
    if status == 1:
        print("CHECK!")
    elif status == 2:
        print(f"CHECKMATE! {cd.invert_color(current_player).to_string()} wins.")
        return -1, current_position, current_player, full_move_counter, halfmove_clock, pgn
    elif status == 3:
        print("STALEMATE! Nobody wins")
        return -1, current_position, current_player, full_move_counter, halfmove_clock, pgn

    # TODO add more endgame conditions
    # elif cd.is_insufficient_material(current_position):
    #     print("INSUFFICIENT MATERIAL!")
    #     break
    # elif cd.is_threefold_repetition(current_position):
    #     print("THREEFOLD REPETITION!")
    #     break
    # elif cd.is_fifty_move_rule(current_position):
    #     print("FIFTY MOVE RULE!")
    #     break

    print("\033[7m")
    print(current_position.display_position(), end="")
    print("\033[0m") # Reset color
    print(current_position.get_fen(current_player, halfmove_clock, full_move_counter, "", ""))
    print(pgn)

    # Determine if it's human's turn
    is_human_turn = (current_player == human_color) and USE_HUMAN
    if is_human_turn:
        quit_game = False
        while True:
            status, move = make_human_turn(current_position, current_player)
            if status == 0:
                break  # Move was made successfully
            elif status == -1:
                quit_game = True
                break
        if quit_game:
            return -1, current_position, current_player, full_move_counter, halfmove_clock, pgn
    else:
        # AI's turn
        print("AI is thinking...")
        if CHOOSE_RANDOMLY:
            move = cd.find_random_move(current_position, current_player)
        else:
            move, debugOut = cd.find_best_move(current_position, current_player, MAX_DEPTH, MAX_TIME, DEBUG)
            if (DEBUG):
                print(debugOut)
        # Convert move to algebraic notation for display
        move_str = f"{Move.to_algebraic(move.from_square)} {Move.to_algebraic(move.to_square)}"
        print(f"\033[41mComputer plays: {move_str} \033[0m")

    # Make the selected move
    current_position.make_move(move)

    # Update PGN
    if current_player == Color.WHITE:
        pgn += f"{full_move_counter}."
    else:
        full_move_counter += 1
    pgn += f"{move.get_pgn()} "

    # Switch to next player
    current_player = cd.invert_color(current_player)
    return 0, current_position, current_player, full_move_counter, halfmove_clock, pgn

current_position = Position()
current_player = Color.WHITE
human_color = Color.WHITE
full_move_counter = 1
halfmove_clock = 0
pgn = ""

# Let player choose color
choice = input("Play as (w)hite or (b)lack? ").strip().lower()
if choice == 'q':
    print("Quitting the game")
    exit(0)
if choice == 'b':
    human_color = Color.BLACK

# Print out who plays who
print()
print(f"You are playing as {(human_color.to_string())}")
print(f"AI is playing as {(cd.invert_color(human_color).to_string())}")
if human_color == Color.WHITE:
    print("You play first")
else:
    print("AI plays first")

# Main game loop
while True:
    result, current_position, current_player, full_move_counter, halfmove_clock, pgn = game_loop(current_position, current_player, human_color, full_move_counter, halfmove_clock, pgn)
    if result == -1:
        break  # Game ended

print("\nGame ended.")
print("PGN:")
print(pgn)
